syntax = "proto3";

package consensus;

service Consensus {
  rpc SayHello (HelloRequest) returns (HelloResponse) {}
  rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse) {}
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse) {}
}

message HelloRequest {
  string name = 1;
}

message HelloResponse {
  string message = 1;
}

message Log {
  int64 logTerm = 1;
  string command = 2;
}

message AppendEntriesRequest {
  int32 term = 1;
  int32 leaderId = 2;
  int64 prevLogIndex = 3;
  int32 prevLogTerm = 4;
  repeated Log entries = 5;
  int64 leaderCommit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
}

message RequestVoteRequest {
  int32 term = 1;
  int32 candidateId = 2;
  int32 lastLogIndex = 3;
  int32 lastLogTerm = 4;
}

message RequestVoteResponse {
  int32 term = 1;
  bool granted = 2;
}


service Request {
  rpc SayHello (HelloRequest) returns (HelloResponse) {}
  rpc AddLog (AddLogRequest) returns (AddLogResponse);
}

message AddLogRequest {
  Log log = 1;
}

message AddLogResponse {
  bool success = 1;
  string response = 2;
}